Shipping Labels M1: UI to refund a shipping label

jaclync opened the pull request:
Fixes #2168 

## Why

After the backend support https://github.com/woocommerce/woocommerce-ios/pull/3294, this PR implements the UI integration to request a refund for a shipping label.

## Changes

- Created `RefundShippingLabelViewController` and view model `RefundShippingLabelViewModel` to display refund information with a CTA to request a refund
- Created `PlainTextSectionHeaderView` for displaying plain text in a table view section header. The reasons I didn't create a xib:
  - The view only contains one `UILabel` and the logic was pretty straightforward
  - There is no option to create a xib when creating a `UITableViewHeaderFooterView`
- In `OrderDetailsViewController`, integrated with `RefundShippingLabelViewController` when the user taps on the refund request CTA

## Testing

Prerequisites: the site has [WooCommerce Shipping & Tax plugin](https://docs.woocommerce.com/document/woocommerce-shipping-and-tax/). You can set up your store to use a test credit card with instructions in p91TBi-3xD-p2.

Please note that in testing mode, a shipping label refund only works if it is requested **shortly** after label creation.

- Go to the orders tab, open an order with at least one physical product
- In wp-admin, create a shipping label for the order above
- Right after the shipping label is created, go back to the app and pull down to refresh the order details --> a shipping label package should be shown after syncing
- Tap on the ellipsis in the package card section header, then tap "Request a Refund" --> a screen should be shown with purchase date and refundable amount, plus a CTA at the bottom
- Tap "Refund Label ({refundable_amount})" CTA --> it should navigate back to order details, then a success snackbar is shown when the refund is requested. In the meantime, the package card should be updated to refunded state (no order items shown)

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/101447118-8ae6f400-395f-11eb-8a43-f57bb86237ea.gif)

portrait | landscape
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-12-08 at 13 46 17](https://user-images.githubusercontent.com/1945542/101447191-ace07680-395f-11eb-82f1-79e263e16a03.png) | ![Simulator Screen Shot - iPhone 11 - 2020-12-08 at 13 46 40](https://user-images.githubusercontent.com/1945542/101447198-b073fd80-395f-11eb-8802-22fb082295b3.png)
![Simulator Screen Shot - iPhone 11 - 2020-12-08 at 13 47 20](https://user-images.githubusercontent.com/1945542/101447200-b10c9400-395f-11eb-8a06-d85832b41f6f.png) | ![Simulator Screen Shot - iPhone 11 - 2020-12-08 at 13 47 24](https://user-images.githubusercontent.com/1945542/101447201-b1a52a80-395f-11eb-83b7-ab5274c44877.png)

Please notice the snackbar at the bottom:

error state | success state
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-12-08 at 13 47 03](https://user-images.githubusercontent.com/1945542/101447256-ce416280-395f-11eb-8424-335011049504.png) | ![Simulator Screen Shot - iPhone 11 - 2020-12-08 at 14 04 53](https://user-images.githubusercontent.com/1945542/101447263-d1d4e980-395f-11eb-861a-7515fca4136f.png)


Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
Create `RefundShippingLabelViewController` and view model to display refund information with a CTA to request a refund.

Garance91540 reviewed (approved):
👏yeah👏!👏

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift:
```diff
@@ -375,8 +375,13 @@ private extension OrderDetailsViewController {
+                self?.navigationController?.popViewController(animated: true)
+            }
+            // Disables the bottom bar (tab bar) when requesting a refund.
+            refundViewController.hidesBottomBarWhenPushed = true
```
👌

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/RefundShippingLabelViewController.swift:
```diff
@@ -0,0 +1,204 @@
+    private let viewModel: RefundShippingLabelViewModel
+    private let rows: [Row]
+    private let noticePresenter: NoticePresenter
+    private let onDismiss: () -> Void
```
What do you think about calling this `onComplete` since the VC has not been dismissed yet.

jaclync replied:
sounds good! updated in https://github.com/woocommerce/woocommerce-ios/pull/3306/commits/9bd8ddd9b7ba9447f064a10049b4bca1cdde5a17

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/RefundShippingLabelViewController.swift:
```diff
@@ -0,0 +1,204 @@
+// MARK: Action Handling
+private extension RefundShippingLabelViewController {
+    func refundShippingLabel() {
+        let action = ShippingLabelAction.refundShippingLabel(shippingLabel: shippingLabel) { result in
+            self.showRefundResultNotice(result: result)
+        }
+        ServiceLocator.stores.dispatch(action)
+        onDismiss()
```
What do you think about moving this to the `ViewModel` layer?
That way we would have a more clear input and output action that will make it easy to unit test. 
Also, we could inject `ServiceLocator.stores` and prevent any kind of "shared state craziness" in tests.

jaclync replied:
yea sounds like a good idea, updated in https://github.com/woocommerce/woocommerce-ios/pull/3306/commits/a4d9c1f7c05a73b0a134a39dec5f545b8af1c38d.

originally I imagined the view model to just provide calculated UI properties for view display, but including the action in the view model does help with testing 😃

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/RefundShippingLabelViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        tableView.delegate = self
+        registerTableViewCellsAndHeader()
+
+        tableView.tableFooterView = UIView()
```
nit: I think we have a helper called `tableView.applyFooterViewForHidingExtraRowPlaceholders` is that what you are trying to achieve?

jaclync replied:
yes thanks for the suggestion! updated in https://github.com/woocommerce/woocommerce-ios/pull/3306/commits/9bd8ddd9b7ba9447f064a10049b4bca1cdde5a17

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/RefundShippingLabelViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        tableView.backgroundColor = .basicBackground
+        tableView.estimatedSectionHeaderHeight = Constants.sectionHeight
+        tableView.sectionHeaderHeight = UITableView.automaticDimension
+        tableView.bounces = false
```
[Q] I don't think I understand the reason for this, could you elaborate a bit? 😇

jaclync replied:
With small-medium font size in portrait mode (probably >= 50% of the use case), this screen doesn't require scrolling and it just feels a bit odd to be able to "scroll" with `bounces` enabled. I can do either way though! Feel free to try it with `bounces` enabled/disabled and see which one feels more intuitive 🙂

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/RefundShippingLabelViewController.swift:
```diff
@@ -0,0 +1,204 @@
+        return cell
+    }
+
+    private func rowAtIndexPath(_ indexPath: IndexPath) -> Row {
```
Do you have any plans to reuse this function elsewhere? If not I don't see a problem with having this statement inline on `cellForRowAtIndexPath`, what do you think?

jaclync replied:
yea there is really no benefit having it in a separate function, updated in https://github.com/woocommerce/woocommerce-ios/pull/3306/commits/9bd8ddd9b7ba9447f064a10049b4bca1cdde5a17

Ecarrion commented on WooCommerce/Classes/ViewRelated/ReusableViews/Section Headers/PlainTextSectionHeaderView.swift:
```diff
@@ -0,0 +1,31 @@
+        label.numberOfLines = 0
+
+        contentView.addSubview(label)
+        contentView.pinSubviewToAllEdges(label, insets: .init(top: 16, left: 16, bottom: 16, right: 16))
```
nit: should we move this to a `Constants` enum?

jaclync replied:
yup! updated in https://github.com/woocommerce/woocommerce-ios/pull/3306/commits/9bd8ddd9b7ba9447f064a10049b4bca1cdde5a17

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/RefundShippingLabelViewModel.swift:
```diff
@@ -0,0 +1,24 @@
+    let refundButtonTitle: String
+
+    init(shippingLabel: ShippingLabel, currencyFormatter: CurrencyFormatter) {
+        self.purchaseDate = shippingLabel.dateCreated.toString(dateStyle: .medium, timeStyle: .short)
+        self.refundableAmount = currencyFormatter.formatAmount(Decimal(shippingLabel.refundableAmount), with: shippingLabel.currency) ?? ""
+        self.refundButtonTitle = String.localizedStringWithFormat(Localization.refundButtonTitleFormat, refundableAmount)
```
Do you think it's worth it to unit test these string formats?

jaclync replied:
yup great point! added in https://github.com/woocommerce/woocommerce-ios/pull/3306/commits/a4d9c1f7c05a73b0a134a39dec5f545b8af1c38d

Ecarrion reviewed (commented):
Code looks good! 🚀 
I just left you a suggestion that could help us when adding unit tests to flow.

Testing also seems to work in all cases! 🎉 
I did notice an issue on bigger fonts that hides some content on the screen. Feel free to address this on a diff PR if your prefer.

<img width="393" alt="Screen Shot 2020-12-10 at 9 36 21 AM" src="https://user-images.githubusercontent.com/562080/101792055-43cd4000-3ad2-11eb-9d3d-5dc3a2a95e56.png">

jaclync committed:
Merge branch 'develop' into issue/2168-refund-shipping-label-ui

jaclync committed:
Improvements based on CR comments.

jaclync committed:
Move refund action from `RefundShippingLabelViewController` to `RefundShippingLabelViewModel` with unit tests.

jaclync commented:
> I did notice an issue on bigger fonts that hides some content on the screen. Feel free to address this on a diff PR if your prefer.

@Ecarrion thank you for testing with large font size! just double checking we are seeing the same issue: the header text is dominating the vertical space, and because it's stuck at the top it leaves very little space to scroll the content below. I added this to a separate issue https://github.com/woocommerce/woocommerce-ios/issues/3334 with design & accessibility enhancements to improve later.

To fix this, I plan to convert the header text from a section header view to a table's `tableHeaderView` or a table view cell so that it scrolls with the content below. Lemme know if you have better suggestions!

Also updated the PR for another look!

Ecarrion reviewed (approved):
Looks great! 🚀
